feat: add `ArgillaSpaCyTransformersTrainer` & improve `ArgillaSpaCyTrainer`

alvarobartt opened the pull request:
# Description

This PR adds support for `spacy-transformers` via the new `ArgillaSpaCyTransformersTrainer` class allowing the user to lock the `transformer` model not to be updated, and the `ArgillaSpaCyTrainer` is improved to allow re-using the `tok2vec` or freeze it, if available.

Besides that, this PR also includes a new arg in `ArgillaTrainer` named `framework_kwargs` which is a Python dict to contain the framework-specific kwargs, in this case intended to be created as `{"update_transformer": False}` and `{"freeze_tok2vec": False}` for both `ArgillaSpaCyTransformersTrainer` and `ArgillaSpaCyTrainer`, respectively. Ideally, we should also move the `spaCy` specific-args already included as `ArgillaTrainer` args.

**Type of change**

- [X] New feature (non-breaking change which adds functionality)
- [X] Improvement (change adding some improvement to an existing functionality)

**How Has This Been Tested**

- [X] Add unit-tests for `ArgillaSpaCyTransformersTrainer`

**Checklist**

- [X] I added relevant documentation
- [X] follows the style guidelines of this project
- [X] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [ ] I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

alvarobartt committed:
docs: add notes for `spaCy` training in notebooks

alvarobartt committed:
feat: add `optimize` arg

alvarobartt committed:
refactor: add `_ArgillaSpaCyTrainerBase`

alvarobartt committed:
feat: add `ArgillaSpaCyTrainer` and `ArgillaSpaCyTransformersTrainer`

alvarobartt committed:
feat: add `Framework.SPACY_TRANSFORMERS`

alvarobartt committed:
test: add `ArgillaSpaCyTransformersTrainer` tests

davidberenstein1957 commented on src/argilla/training/base.py:
```diff
@@ -46,6 +46,7 @@ def __init__(
         train_size: Optional[float] = None,
         seed: Optional[int] = None,
         gpu_id: Optional[int] = -1,
+        framework_kwargs: Optional[dict] = {},
```
@alvarobartt For me it makes more sense to add this to the `update_config()`.

alvarobartt replied:
As we can also define the `optimize` value there I'd keep it as is, plus in an upcoming iteration moving the `spaCy`-specific stuff we have there to the `framework_kwargs` arg, to be provided in the `__init__`

davidberenstein1957 commented on src/argilla/training/spacy.py:
```diff
@@ -35,18 +37,17 @@ def __init__(
         language: Optional[str] = None,
         gpu_id: Optional[int] = -1,
         model: Optional[str] = None,
+        optimize: Literal["efficiency", "accuracy"] = "efficiency",
```
perhaps for me this makes sense to add to `update_config`too

alvarobartt replied:
It's not so straight forward to modify that from the configuration file itself, so we should keep it as is, since we're using the `optimize` value to generate one config or the other initially, more things differ in those configuration files, not just a string as it may seem according to the `optimize` arg

davidberenstein1957 commented on tests/training/test_spacy.py:
```diff
@@ -13,89 +13,98 @@
+from tests.training.helpers import cleanup_spacy_config, train_with_cleanup
+
+
+@pytest.mark.parametrize(
```
You are not really passing the new variables in the tests right?

alvarobartt replied:
Yes we are, as we included the args in the `test_` methods and then created the `TestSpaCyTrainer` class, so those values are passed to every function, and we're "looping" twice, once for `spacy` and another time for `spacy-transformers` 😄

davidberenstein1957 reviewed (commented):
LGTM and thanks for the fix. Perhaps it makes more sense to me to pass the variable via the `update_config` to align with the usage of the rest of the frameworks. Also, I don't see the new config options being test right?

alvarobartt committed:
Merge branch 'develop' into feat/spacy-and-spacy-transformers

alvarobartt committed:
fix(test): add missing `self` in `TestSpaCyTrainer` methods

alvarobartt committed:
fix: don't check `Enum` against `value`

alvarobartt committed:
fix: set flag before `super().__init__`

alvarobartt committed:
fix: add `require_version(spacy>=3.5.3)`

Otherwise the configuration for `spacy-transformers` generated via `init_config` doesn't support it

alvarobartt committed:
fix: `tok2vec` freezing

davidberenstein1957 commented on src/argilla/training/spacy.py:
```diff
@@ -95,9 +100,10 @@ def __init__(
         )
         self._train_dataset_path = "./train.spacy"
-        self._eval_dataset_path = "./dev.spacy" if self._eval_dataset else None
+        self._eval_dataset_path = "./dev.spacy" if self._eval_dataset else "./train.spacy"
```
do we need to provide an eval dataset by default? For me it feels weird to use the training dataset to do this.

alvarobartt replied:
Yes, it is mandatory, the solution/workaround here is to use the train dataset, but note that here I just simplified the code, before we had the exact same thing just that we were setting it to None first and then setting it to `train` if it was None

But as it's mandatory I also agree with raising an error if not provided

davidberenstein1957 replied:
No, this seems like a better solution. Thanks

alvarobartt committed:
chore: upgrade `spacy` and add `spacy-transformers`

alvarobartt committed:
test: skip `test_predict_wo_training` for `bert-tiny`

davidberenstein1957 committed:
chore: added integration with Feedback task

davidberenstein1957 committed:
Merge branch 'develop' into feat/spacy-and-spacy-transformers

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci
